### PR TITLE
Throw an error if you try to save a Secure Variable with no items

### LINKED ISSUE #13393: Secure Variables: compact/remove empty Items[] on save
Currently, an added-but-empty row will append to the Items array, but we should clear these out on save.

image
image

### PR DESCRIPTION
**Pull request by philrenaud**
Resolves #13393

1. Compacts items (key/values) by filtering out those with no key/value
2. Throws an error message to the user if they attempt to save a Secure Variable with no items

### CODE REVIEW

File: ui/tests/integration/components/variable-paths-test.js [hunk 1]
```diff
@@ -77,21 +77,21 @@ module('Integration | Component | variable-paths', function (hooks) {
       .dom('tbody tr:first-child td:first-child a')
       .hasAttribute(
         'href',
-        '/ui/variables/foo/bar/baz',
+        '/ui/variables/var/foo/bar/baz',
```
Comment (philrenaud) [inline ui/tests/integration/components/variable-paths-test.js:80]: Side-effects: correcting for a previous change to add /var to disambiguate variable path URLs

File: ui/app/components/secure-variable-form.js [hunk 2]
```diff
@@ -77,9 +77,19 @@ export default class SecureVariableFormComponent extends Component {
   async save(e) {
     e.preventDefault();
     try {
+      const nonEmptyItems = this.keyValues.filter(
+        (item) => item.key && item.value
```
Comment (ChaiWithJai) [inline ui/app/components/secure-variable-form.js:81]: `nit`: This doesn't factor for whitespace. I set the key to be ` ` and the value to be ` `. And this is still valid. Maybe add `trim`?

File: ui/app/models/variable.js [hunk 3]
```diff
@@ -67,6 +67,6 @@ export default class VariableModel extends Model {
    */
   setAndTrimPath() {
     this.set('path', trimPath([this.path]));
-    this.id = this.path;
+    this.set('id', this.get('path'));
```
Comment (ChaiWithJai) [inline ui/app/models/variable.js:70]: `nit` `non-blocking`:  you could have this be `setProperties`.

https://api.emberjs.com/ember/4.4/functions/@ember%2Fobject/setProperties

File: ui/mirage/config.js [hunk 4]
```diff
@@ -848,13 +848,13 @@ export default function () {
 
   this.put('/var/:id', function (schema, request) {
     const { Path, Namespace, Items } = JSON.parse(request.requestBody);
-    server.create('variable', {
+    const newVar = server.create('variable', {
```
Comment (ChaiWithJai) [inline ui/mirage/config.js:851]: `nit` `non-blocking`: you could just `return` here if you wanna play code golf.

File: ui/tests/acceptance/secure-variables-test.js [hunk 5]
```diff
@@ -119,6 +119,25 @@ module('Acceptance | secure variables', function (hooks) {
     assert.notOk(fooLink, 'foo0 file is no longer present');
   });
 
+  test('it does not allow you to save if you lack Items', async function (assert) {
+    defaultScenario(server);
```
Comment (ChaiWithJai) [inline ui/tests/acceptance/secure-variables-test.js:123]: `question`: Should this move into a `beforeEach` hook?

File: ui/tests/acceptance/secure-variables-test.js [hunk 6]
```diff
@@ -119,6 +119,25 @@ module('Acceptance | secure variables', function (hooks) {
...
+    window.localStorage.nomadTokenSecret = server.db.tokens[0].secretId;
```
Comment (ChaiWithJai) [inline ui/tests/acceptance/secure-variables-test.js:124]: `question`: Should we clean this up at the end of the test?

**Review by ChaiWithJai — APPROVED** [review]
LGTM some small comments. I do think that logic for actions in `secure-variable-form` is starting to get exceptionally complex.

File: ui/app/components/secure-variable-form.js [hunk 7]
```diff
@@ -77,9 +77,19 @@ export default class SecureVariableFormComponent extends Component {
...
+      );
+      if (!nonEmptyItems.length) {
+        throw new Error('Please provide at least one key/value pair.');
```
[the remark this answers is not in the document]
Comment (philrenaud) [reply ui/app/components/secure-variable-form.js:84]: no timeout on error message is deliberate; we want the user to acknowledge that a mistake was made.

File: ui/app/components/secure-variable-form.js (hunk 2 above)
Comment (philrenaud) [reply ui/app/components/secure-variable-form.js:81]: Hm, I think this is probably safe to do for the key, but I'm going to avoid it for value.

I think there's a world where someone wants a value of a whitespace character that would otherwise be trimmed for their secure variable, and we shouldn't necessarily nuke that. 

cc @angrycub  / @tgross in case there's a good reason to not trim keys as well.

File: ui/app/models/variable.js (hunk 3 above)
Comment (philrenaud) [reply ui/app/models/variable.js:70]: I think that would cause an error; I want to set the `id` only after I set the `path` above, and editing both in a single `setProperties` would not give me a modified path.

File: ui/tests/acceptance/secure-variables-test.js (hunk 5 above)
Comment (philrenaud) [reply ui/tests/acceptance/secure-variables-test.js:123]: Hm not yet; I imagine there will be scenarios for this form where I don't want defaultScenario(), but set up some something custom / edge-casey.

File: ui/tests/acceptance/secure-variables-test.js (hunk 6 above)
Comment (philrenaud) [reply ui/tests/acceptance/secure-variables-test.js:124]: I think this probably isn't harmful / no need to clean it up yet. The precedent for this is found in another dozen+ test files throughout.

File: ui/app/models/variable.js (hunk 3 above)
Comment (ChaiWithJai) [reply ui/app/models/variable.js:70]: Gotcha! We should make sure `setAndTrimPath` doesn't run on `edit` and only on `create` since `path` shouldn't be a mutable property.
